### PR TITLE
Dev captcha v2

### PR DESCRIPTION
**Pull request by alexbpx**

### CODE REVIEW

File: src/main/java/com/perimeterx/api/PerimeterX.java [hunk 1]
```diff
@@ -166,32 +167,11 @@ public PXContext pxVerify(HttpServletRequest req, HttpServletResponseWrapper res
                 context.setFirstPartyRequest(true);
                 return context;
             }
-
-            // Remove captcha cookie to prevent re-use
             Cookie cookie = new Cookie(Constants.COOKIE_CAPTCHA_KEY, StringUtils.EMPTY);
```
Comment (yaronpx) [inline src/main/java/com/perimeterx/api/PerimeterX.java]: this should also be removed

File: src/main/java/com/perimeterx/api/blockhandler/DefaultBlockHandler.java [hunk 2]
```diff
@@ -24,41 +24,65 @@
...
     public void handleBlocking(PXContext context, PXConfiguration pxConfig, HttpServletResponseWrapper responseWrapper) throws PXException {
-        String blockPage = getPage(context, pxConfig);
-        String action;
+        Map<String, String> props = new HashMap<>();
+        String filePrefix = null;
+        String blockPageResponse = null;
+        if (context.getBlockAction().equals(BlockAction.CHALLENGE) && context.getBlockActionData() != null) {
+            blockPageResponse= context.getBlockActionData();
+        }
+        if (context.getBlockAction().equals(BlockAction.CAPTCHA)) {
```
Comment (yaronpx) [inline src/main/java/com/perimeterx/api/blockhandler/DefaultBlockHandler.java]: what about block action?

File: src/main/java/com/perimeterx/api/blockhandler/templates/TemplateFactory.java [hunk 3]
```diff
@@ -53,10 +47,24 @@ public static String getTemplate(PXContext pxContext, PXConfiguration pxConfig,
...
+        String urlVid = pxContext.getVid() != null ? pxContext.getVid() : "";
+        if (pxConfig.isFirstPartyEnabled() && !pxContext.isMobileToken()){
+            String prefix = pxConfig.getAppId().substring(2);
+            blockScript = "/" + prefix + Constants.FIRST_PARTY_CAPTCHA_PATH + "/captcha.js?a=" + pxContext.getBlockAction().getCode() + "&u=" + pxContext.getUuid() + "&v=" + urlVid + "&m=" + (pxContext.isMobileToken() ? "1" :"0");
+            jsClientSrc = "/" + prefix + Constants.FIRST_PARTY_VENDOR_PATH;
+            hostUrl = "/" + prefix + Constants.FIRST_PARTY_XHR_PATH;
+        }
+        else{
+            blockScript = "//" + Constants.CAPTCHA_HOST + "/" + pxConfig.getAppId() + "/captcha.js?a=" + pxContext.getBlockAction().getCode() + "&u=" + pxContext.getUuid() + "&v=" + urlVid + "&m=" + (pxContext.isMobileToken() ? "1" :"0");
+            jsClientSrc = "//" + Constants.CLIENT_HOST + "/" + pxConfig.getAppId() + "/main.min.js";
```
Comment (yaronpx) [inline src/main/java/com/perimeterx/api/blockhandler/templates/TemplateFactory.java]: should add first party `hostUrl` as well here

File: src/main/java/com/perimeterx/api/PerimeterX.java (hunk 1 above)
Comment (alexbpx) [reply src/main/java/com/perimeterx/api/PerimeterX.java]: I don't see the comment in the code

File: src/main/java/com/perimeterx/api/blockhandler/templates/TemplateFactory.java (hunk 3 above)
Comment (alexbpx) [reply src/main/java/com/perimeterx/api/blockhandler/templates/TemplateFactory.java]: why, if it's not the firstparty flow?

Comment (yaronpx) [reply src/main/java/com/perimeterx/api/blockhandler/templates/TemplateFactory.java]: sorry, I meant third party. now I see you already set it a few lines above

**Comment by yaronpx** [comment]
please fix all conflicts (merge dev into your branch)

File: src/main/java/com/perimeterx/api/blockhandler/templates/TemplateFactory.java [hunk 4]
```diff
@@ -53,10 +47,25 @@ public static String getTemplate(PXContext pxContext, PXConfiguration pxConfig,
...
+        String hostUrl = pxContext.getCollectorURL();
+        if (pxConfig.isFirstPartyEnabled() && !pxContext.isMobileToken()){
+            String prefix = pxConfig.getAppId().substring(2);
+            blockScript = "/" + prefix + Constants.FIRST_PARTY_CAPTCHA_PATH + "/captcha.js?a=" + pxContext.getBlockAction().getCode() + "&u=" + pxContext.getUuid() + "&v=" + urlVid + "&m=" + (pxContext.isMobileToken() ? "1" :"0");
+            jsClientSrc = "/" + prefix + Constants.FIRST_PARTY_VENDOR_PATH;
+            hostUrl = "/" + prefix + Constants.FIRST_PARTY_XHR_PATH;
+        }
+        String blockAction = pxContext.getBlockAction().getCode();
+        if ("b".equals(blockAction)){
+            blockScript = "";
```
Comment (yaronpx) [inline src/main/java/com/perimeterx/api/blockhandler/templates/TemplateFactory.java]: this is incorrect. if block action is `b`, then the query parameter `a` should be `b`. but the block page should still load the captcha script

File: src/test/java/com/perimeterx/api/TemplatesTest.java [hunk 5]
```diff
@@ -65,24 +68,13 @@ public void testDefaultPageContainsLogo() throws Exception {
...
                 .build();
         pxContext.setBlockAction("b");
-        String actualHTML = TemplateFactory.getTemplate(pxContext,pxConfig,"block.mustache");
+        Map<String, String> props = TemplateFactory.getProps(pxContext, pxConfig);
+        String actualHTML = TemplateFactory.getTemplate("captcha_template.mustache", props);
         assertTrue( actualHTML.contains("http://www.google.com/logo.jpg") );
     }
 
 
-    @Test
```
Comment (yaronpx) [inline src/test/java/com/perimeterx/api/TemplatesTest.java]: why did you remove this test? it's still important that we'll be able to inject custom css to the template

File: src/test/java/com/perimeterx/api/TemplatesTest.java [hunk 6]
```diff
@@ -94,23 +86,11 @@ public void testDefualtPageContainsCSS() throws Exception {
...
                 .build();
         pxContext.setBlockAction("b");
-        String actualHTML = TemplateFactory.getTemplate(pxContext,pxConfig,"block.mustache");
+        Map<String, String> props = TemplateFactory.getProps(pxContext, pxConfig);
+        String actualHTML = TemplateFactory.getTemplate("captcha_template.mustache", props);
         assertTrue( actualHTML.contains("http://www.google.com/stylesheet.css") );
     }
 
-    @Test
-    public void testDefualtPageContainsJS() throws Exception {
```
Comment (yaronpx) [inline src/test/java/com/perimeterx/api/TemplatesTest.java]: same as above
